lib/matplotlib/mathtext.pyi

```diff
@@ -14,13 +14,8 @@ from matplotlib.typing import ColorType
 class MathTextParser:
     def __init__(self, output: Literal["path", "agg", "raster", "macosx"]) -> None: ...
     def parse(
-        self,
-        s: str,
-        dpi: float = ...,
-        prop: FontProperties | None = ...,
-        *,
-        antialiased: bool | None = ...
-    ): ...
+        self, s: str, dpi: float = ..., prop: FontProperties | None = ..., *, antialiased: bool | None = ...
+    ) -> RasterParse | VectorParse: ...
```

Comment:
Wish there was a better way here... a _given_ MathTextParser knows which it will return, but that is a state of the mathtext parser, not of this method...

Could maybe do some funky generic things with `__new__`? (I'd say `__init__`, but don't have access to the type of the constructed object there) but that gets weird quickly, so probably not worth it.

Reply:
Something like this seems to work if we want to apply it:
```patch
diff --git a/lib/matplotlib/mathtext.pyi b/lib/matplotlib/mathtext.pyi
index 706f83a6b1..1b0468d59a 100644
--- a/lib/matplotlib/mathtext.pyi
+++ b/lib/matplotlib/mathtext.pyi
@@ -1,5 +1,8 @@
 import os
+from typing import Generic, IO, Literal, TypeVar, overload
+
 from matplotlib.font_manager import FontProperties
+from matplotlib.typing import ColorType

 # Re-exported API from _mathtext.
 from ._mathtext import (
@@ -8,14 +11,17 @@ from ._mathtext import (
     get_unicode_index as get_unicode_index,
 )

-from typing import IO, Literal
-from matplotlib.typing import ColorType
+_ParseType = TypeVar("_ParseType", RasterParse, VectorParse)

-class MathTextParser:
-    def __init__(self, output: Literal["path", "agg", "raster", "macosx"]) -> None: ...
+class MathTextParser(Generic[_ParseType]):
+    @overload
+    def __init__(self: MathTextParser[VectorParse], output: Literal["path"]) -> None: ...
+    @overload
+    def __init__(self: MathTextParser[RasterParse], output: Literal["agg", "raster", "macosx"]) -> None: ...
+    def __init__(self: MathTextParser[_ParseType], output: Literal["path", "agg", "raster", "macosx"]) -> None: ...
     def parse(
         self, s: str, dpi: float = ..., prop: FontProperties | None = ..., *, antialiased: bool | None = ...
-    ) -> RasterParse | VectorParse: ...
+    ) -> _ParseType: ...

 def math_to_image(
     s: str,
```

Reply:
Though it seems like we'd need to add extra casts on usage to specify the generic, so it's maybe not so great?